### PR TITLE
Fix some potential `ChainState` issues

### PR DESCRIPTION
**Pull request by afck**
## Motivation

The admin ID can't change, and we are locking the `ChainState` each time we access it.

We are locking and unlocking the `ChainState` multiple times in some cases to obtain the components of a chain's tip state (block hash, height and timestamp). If the state changed in between those instances, the components won't refer to the same state.

In some cases we return an error if the local node's chain is ahead of the `ChainState`. But in multi-owner chains, this can happen legitimately.

## Proposal

Move `admin_id` out of `ChainState`. Fetch the tip state in one go. Only do the check for single-owner chains.

Also, don't unnecessarily collect the known keys' owners, and use `contains_key` instead.

## Test Plan

CI should catch regressions.

## Release Plan

- Nothing to do / These changes follow the usual release cycle.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-core/src/client/mod.rs [hunk 1]
```diff
@@ -610,6 +601,16 @@ impl<P: 'static, S: Storage> ChainClient<P, S> {
         Unsend::new(self.state().inner.map(|state| state.pending_blobs()))
     }
 }
+#[cfg(with_testing)]
```
Comment (jvff) [inline linera-core/src/client/mod.rs:604]: Nit: adding an empty line to separate the items

```suggestion

#[cfg(with_testing)]
```

File: linera-core/src/client/mod.rs [hunk 2]
```diff
@@ -675,21 +677,23 @@ where
     /// local chain.
     #[tracing::instrument(level = "trace")]
     async fn pending_message_bundles(&self) -> Result<Vec<IncomingBundle>, ChainClientError> {
-        if self.next_block_height() != BlockHeight::ZERO && self.options.message_policy.is_ignore()
-        {
-            return Ok(Vec::new()); // OpenChain is already received, other are ignored.
-        }
```
Comment (deuszx) [inline linera-core/src/client/mod.rs:681]: Why did you move it, was it wrong?

**Review by deuszx — APPROVED** [review]

File: linera-core/src/client/mod.rs (hunk 2 above)
Comment (afck) [reply linera-core/src/client/mod.rs:681]: Our `ChainState`'s `next_block_height` could be out of sync with the `local_node`. We should use the latest information from the local node to make decisions.

Comment (deuszx) [reply linera-core/src/client/mod.rs:681]: The only part that matters here is whether `next_block_height` is non-ZERO - in which case we should return `Ok(Vec::new())` – this is only the case when a chain makes THE FIRST the new block before last update of `self.next_block_height` and the query. Does that sound correct?

Comment (afck) [reply linera-core/src/client/mod.rs:681]: Yes.
So here it could happen that the chain state is stale and thinks we're still at block height 0, but the chain info knows we're already at block height 1. In that case it makes sense to return `Vec::new()`.

File: linera-core/src/client/mod.rs [hunk 3]
```diff
@@ -675,21 +677,23 @@ where
...
             .await?
             .info;
-        ensure!(
-            info.next_block_height == self.next_block_height(),
-            ChainClientError::WalletSynchronizationError
-        );
+        {
+            let state = self.state();
+            ensure!(
+                state.has_other_owners(&info.manager.ownership) || info.tip() == state.tip(),
```
Comment (jvff) [inline linera-core/src/client/mod.rs:690]: Question: I'm not sure I understand why this new condition was added :sweat_smile:

File: linera-core/src/client/mod.rs [hunk 4]
```diff
@@ -885,13 +889,11 @@ where
         if info.next_block_height == next_block_height {
             // Check that our local node has the expected block hash.
             ensure!(
-                self.block_hash() == info.block_hash,
+                (info.block_hash, info.timestamp) == (block_hash, tip_time),
```
[the remark this answers is not in the document]
Comment (afck) [reply linera-core/src/client/mod.rs:892]: Good point, that's redundant. I can revert that line.

File: linera-core/src/client/mod.rs (hunk 3 above)
Comment (afck) [reply linera-core/src/client/mod.rs:690]: Because if the chain has other owners it could have been extended without us knowing about it. So our chain state we read from our wallet not matching the actual state is not an error.

File: linera-core/src/client/mod.rs (hunk 2 above)
Comment (deuszx) [reply linera-core/src/client/mod.rs:681]: We update `self.next_block_height` in a couple of places so it's not "stale forever".

**Review by jvff — APPROVED** [review]

File: linera-core/src/client/mod.rs [hunk 5]
```diff
@@ -610,6 +601,16 @@ impl<P: 'static, S: Storage> ChainClient<P, S> {
         Unsend::new(self.state().inner.map(|state| state.pending_blobs()))
     }
 }
+#[cfg(with_testing)]
+impl<P: 'static, S: Storage> ChainClient<P, S> {
+    pub fn block_hash(&self) -> Option<CryptoHash> {
+        self.state().tip().1
```
Comment (ma2bd) [inline linera-core/src/client/mod.rs:607]: This `.1` is a little bit of a regression. Can we keep the individual getters and can we define a `struct Tip` with proper field names?

File: linera-core/src/client/mod.rs (hunk 3 above)
Comment (ma2bd) [reply linera-core/src/client/mod.rs:695]: Seems like but we may just want a warning. The chain could have other owners now unbeknownst to us.

File: linera-core/src/client/mod.rs [hunk 6]
```diff
@@ -1736,25 +1736,22 @@ where
...
                         ChainClientError::BlockProposalError(
                             "Client state has a different pending block; \
-                             use the `linera retry-pending-block` command to commit that first"
+                            use the `linera retry-pending-block` command to commit that first"
                         )
                     );
                 }
             }
-        }
+            state.tip()
```
Comment (ma2bd) [inline linera-core/src/client/mod.rs:1753]: Definitely create a data structure so that we can write `tip.timestamp` instead of `tip_time` etc

File: linera-core/src/client/mod.rs [hunk 7]
```diff
@@ -1979,14 +1973,18 @@ where
     /// This will usually be the current time according to the local clock, but may be slightly
     /// ahead to make sure it's not earlier than the incoming messages or the previous block.
     #[tracing::instrument(level = "trace", skip(incoming_bundles))]
-    async fn next_timestamp(&self, incoming_bundles: &[IncomingBundle]) -> Timestamp {
+    async fn next_timestamp(
+        &self,
+        incoming_bundles: &[IncomingBundle],
+        tip_time: Timestamp,
```
Comment (ma2bd) [inline linera-core/src/client/mod.rs:1979]: Can we read `self.state().tip().timestamp` ?

File: linera-core/src/data_types.rs [hunk 8]
```diff
@@ -165,6 +165,12 @@ pub struct ChainInfo {
     pub requested_received_log: Vec<ChainAndHeight>,
 }
 
+impl ChainInfo {
+    pub fn tip(&self) -> (BlockHeight, Option<CryptoHash>, Timestamp) {
+        (self.next_block_height, self.block_hash, self.timestamp)
```
Comment (ma2bd) [inline linera-core/src/data_types.rs:170]: `Tip { ... }`

**Review by ma2bd — COMMENTED** [review]
The PR is doing several things (removing a lock, fixing some issue?) and as a result it's not easy to read

File: linera-core/src/client/mod.rs (hunk 3 above)
Comment (deuszx) [reply linera-core/src/client/mod.rs:695]: I'd imagine we need to learn (sync) the chain owners before we process its blocks?

Comment (afck) [reply linera-core/src/client/mod.rs:695]: That's true, this would still fail if we deleted our storage.

(At least if we don't, it can't: If at the highest block known to us we are still the only owners, then ownership can't have changed.)

**Comment by afck** [comment]
Closing in favor of https://github.com/linera-io/linera-protocol/pull/2675.

Thanks for the comments, and thanks to @andresilva91 for fixing the underlying issue that caused the test to fail! #2675 should pass CI now and be less controversial.
